animation_nodes/algorithms/mesh/triangulate_mesh.pyx

```diff
@@ -0,0 +1,316 @@
+            newIndices[polyIndex + 1] = verticesData[earIndex].index
+            newIndices[polyIndex + 2] = verticesData[verticesData[earIndex].next].index
+            polyIndex += 3
+
+            removeEarVertex(polyLength, verticesData, earIndex)
+
+        PyMem_Free(verticesData)
+
+    return newPolygons
+
+# Remove ear-vertex and update neighbor-vertices.
+cdef void removeEarVertex(Py_ssize_t polyLength, Vertex* verticesData, Py_ssize_t earIndex):
+    cdef Py_ssize_t previousIndex, nextIndex
+    previousIndex = verticesData[earIndex].previous
+    nextIndex = verticesData[earIndex].next
+
+    verticesData[earIndex].isEar = False
+    verticesData[nextIndex].previous = previousIndex
+    verticesData[previousIndex].next = nextIndex
+
+    cdef float angle
+    angle = calculateAngleWithSign(verticesData, previousIndex)
+    verticesData[previousIndex].angle = angle
+    if angle >= 0.0 and not pointsInTriangle(polyLength, verticesData, previousIndex):
+        verticesData[previousIndex].isEar = True
+    else:
+        verticesData[previousIndex].isEar = False
+
+    angle = calculateAngleWithSign(verticesData, nextIndex)
+    verticesData[nextIndex].angle = angle
+    if angle >= 0.0 and not pointsInTriangle(polyLength, verticesData, nextIndex):
+        verticesData[nextIndex].isEar = True
+    else:
+        verticesData[nextIndex].isEar = False
+
+cdef struct Vertex:
+    Py_ssize_t index
+    Vector3 location
+    float angle
+    bint isEar
+    Py_ssize_t previous
+    Py_ssize_t next
+
+# Make sure normal (+z) is correct.
+@cython.cdivision(True)
+cdef bint polyCounterClockwise(Vertex* verticesData):
+    cdef Vector3 v1, v2
+    cdef float area = 0.0
+    cdef Py_ssize_t headIndex = 0
+    cdef Py_ssize_t currentIndex
+
+    currentIndex = headIndex
+    while(True):
+        v1 = verticesData[currentIndex].location
+        currentIndex = verticesData[currentIndex].next
+        v2 = verticesData[currentIndex].location
+        area += (v2.x - v1.x) * (v2.y + v1.y)
+        if currentIndex == headIndex: break
+
+    if area > 0.0: return False
+    return True
+
+# Calculate inner angle.
+cdef float calculateAngleWithSign(Vertex* verticesData, Py_ssize_t index):
+    cdef Vector3 v1 = verticesData[verticesData[index].previous].location
+    cdef Vector3 v2 = verticesData[index].location
+    cdef Vector3 v3 = verticesData[verticesData[index].next].location
+    cdef Vector3 ab, bc
+    cdef float angle
+
+    subVec3(&ab, &v2, &v1)
+    subVec3(&bc, &v2, &v3)
+
+    normalizeVec3_InPlace(&ab)
+    normalizeVec3_InPlace(&bc)
+
+    angle = angleNormalizedVec3(&ab, &bc)
+
+    # Make sure normal (+z) is correct, and the polygon is counter clockwise. Convex vertex has sign >= 0.
+    cdef float sign = (v2.x - v1.x) * (v3.y - v1.y) - (v2.y - v1.y) * (v3.x - v1.x)
+    if sign < 0.0: return -(angle - angle % 0.001)
+    return (angle - angle % 0.001)
```

Comment:
Why the need for this `angle - angle % 0.001`?

Reply:
I added this to handle the precision problem. It mainly happens when angles are equal to each other e.g., square polygon.

Reply:
Can you elaborate? What happens if the angles are equal?

Reply:
I have tried without that, now everything is fine.

However, before the angles were changing after some decimal places. May be due to the `pi` that I used before.

Reply:
It is happening when the input mesh is animated.

Reply:
Here is an example:
![Peek 2020-06-18 19-46](https://user-images.githubusercontent.com/30294746/85031520-66724200-b19c-11ea-9caf-d25659a011e6.gif)

Reply:
I see. Do we need `@cython.cdivision(True)` here since `%` is used?

Reply:
I think no. `cython -a` shows white color for those lines.
![Screenshot from 2020-06-18 19-58-27](https://user-images.githubusercontent.com/30294746/85033077-27dd8700-b19e-11ea-96e4-f3265cbc2283.png)